torch/csrc/jit/codegen/cuda/lower_warp_reduce.cpp

```diff
@@ -0,0 +1,546 @@
+  c10::optional<kir::TensorIndex*> findMaybeReplacedTensorIndex(
+      kir::TensorIndex* tensor_index) {
+    auto kir_tv = tensor_index->view();
+    auto tensor_index_it = running_tv_replacement_map_.find(kir_tv);
+    if (tensor_index_it != running_tv_replacement_map_.end()) {
+      return tensor_index_it->second;
+    }
+    return c10::nullopt;
+  }
+
+  //! Iteratve backwards on the currently visible loop scopes
+  //!  and find the first allocate corresponding to the
+  //!  given tv.
+  kir::Allocate* getActiveAllocateFor(kir::TensorView* tv) {
+    for (auto frame_it = running_visible_allocation_stack_.rbegin();
+         frame_it != running_visible_allocation_stack_.rend();
+         frame_it++) {
+      for (auto allocate_it = (*frame_it)->rbegin();
+           allocate_it != (*frame_it)->rend();
+           allocate_it++) {
+        auto candidate_allocate = *allocate_it;
+        if (candidate_allocate->buffer() == tv) {
+          return candidate_allocate;
+        }
+      }
+    }
+    TORCH_INTERNAL_ASSERT(
+        false, "lower_warp_reduce: cannot find allocate for this op");
+    return nullptr;
+  }
+
+  Expr* getFuserTVExpr(kir::Expr* expr) {
+    auto out = expr->outputs()[0];
+    auto out_ti = dynamic_cast<kir::TensorIndex*>(out);
+    if (!out_ti) {
+      return nullptr;
+    }
+    return out_ti->view()->fuserTv()->definition();
+  }
+
+  bool isOpInputRegisterTV(kir::Expr* expr) {
+    auto fuser_expr = getFuserTVExpr(expr);
+    if (!fuser_expr) {
+      return false;
+    }
+
+    for (auto inp : fuser_expr->inputs()) {
+      if (auto inp_tv = dynamic_cast<TensorView*>(inp)) {
+        if (inp_tv->getMemoryType() != MemoryType::Local) {
+          return false;
+        }
+      }
+    }
+
+    return true;
+  }
+
+  bool isOpOutputRegisterTV(kir::Expr* expr) {
+    auto fuser_expr = getFuserTVExpr(expr);
+    if (!fuser_expr) {
+      return false;
+    }
+
+    for (auto out : fuser_expr->outputs()) {
+      if (auto out_tv = dynamic_cast<TensorView*>(out)) {
+        if (out_tv->getMemoryType() != MemoryType::Local) {
+          return false;
+        }
+      }
+    }
+
+    return true;
+  }
+
+  //! Updates map of serially visible reduction tvs, see comment on
+  //!  running_kir_tv_to_allocate_map_.
+  void handle(kir::ReductionOp* reduction) {
+    if (!isOpOutputRegisterTV(reduction)) {
+      return;
+    }
+    auto reduction_ti = dynamic_cast<kir::TensorIndex*>(reduction->out());
+    TORCH_INTERNAL_ASSERT(
+        reduction_ti, "lower_warp_reduce: need to run after indexing");
+
+    // keep track of which reduction buffer this expr writes into
+    auto reduction_allocate = getActiveAllocateFor(reduction_ti->view());
+    running_kir_tv_to_allocate_map_.back()->operator[](reduction_ti->view()) =
+        reduction_allocate;
+  }
+
+  void handle(kir::BroadcastOp* broadcast) {
+    if (!isOpOutputRegisterTV(broadcast) || !isOpOutputRegisterTV(broadcast)) {
```

Comment:
I guess this should be `!isOpInputRegisterTV(...) || `!isOpOutputRegisterTV(...)`.

Reply:
Yes good catch 👍  Fixed, thanks!